Modules/_datetimemodule.c

```diff
@@ -1939,8 +1944,46 @@ wrap_strftime(PyObject *object, PyObject *format, PyObject *timetuple,
             ptoappend = PyBytes_AS_STRING(freplacement);
             ntoappend = PyBytes_GET_SIZE(freplacement);
         }
+#ifdef Py_NORMALIZE_CENTURY
+        else if (ch == 'Y' || ch == 'G') {
+            /* 0-pad year with century as necessary */
+            PyObject *year_str;                     /* py string, year */
+            PyObject *year;                         /* py int, year */
+            long year_long;                         /* year as long int */
+            char year_formatted[SIZEOF_LONG*5/2+2]; /* formatted year for %Y/%G */
+
+            year_long = PyLong_AsLong(PyTuple_GET_ITEM(timetuple, 0));
+            if (year_long == -1 && PyErr_Occurred()) {
+                goto Done;
+            }
+            if (year_long > 1000) {
+                goto PassThrough;
+            }
+            if (ch == 'G') {
+                year_str = PyObject_CallFunction(strftime, "sO", "%G", timetuple);
+                if (year_str == NULL) {
+                    goto Done;
+                }
+                year = PyNumber_Long(year_str);
+                Py_DECREF(year_str);
+                if (year == NULL) {
+                    goto Done;
+                }
+                year_long = PyLong_AsLong(year);
+                Py_DECREF(year);
+                if (year_long == -1 && PyErr_Occurred()) {
+                    goto Done;
+                }
+            }
+            ntoappend = sprintf(year_formatted, "%04ld", year_long);
```

Comment:
Also, let's not introduce more `sprintf`s; use `PyOS_snprintf`. See also:

- https://github.com/python/cpython/issues/101703

Reply:
Ah good to know. Thanks!